core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -131,6 +132,9 @@ private FileIO io(String location) {
     String impl = implFromLocation(location);
     FileIO io = ioInstances.get(impl);
     if (io != null) {
+      if (io instanceof HadoopFileIO && ((HadoopFileIO) io).conf() == null) {
```

Comment:
I've extracted solving this into https://github.com/apache/iceberg/pull/8270. There's an issue with Kryo + deserializing `ResolvingFileIO`, because after de-serializing, the Hadoop conf isn't re-applied.